src/cider/common/interpreters/AggregationHashTable.h

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ * Copyright (c) 2016-2022 ClickHouse, Inc.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <common/Arena.h>
+#include <common/hashtable/FixedHashMap.h>
+#include <common/hashtable/HashMap.h>
+#include <type/data/sqltypes.h>
+
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <type_traits>
+#include <unordered_set>
+
+#include "cider/CiderException.h"
+
+namespace cider {
+
+using AggregateDataPtr = int8_t*;
+
+// It's a wrap for null key, "base" should be one of the HashTables.
+template <typename Base>
+struct AggregatedHashTableWithNullKey : public Base {
+  using Base::Base;
+
+  bool& hasNullKeyData() { return has_null_key_data_; }
+  AggregateDataPtr& getNullKeyData() { return null_key_data_; }
+  bool hasNullKeyData() const { return has_null_key_data_; }
+  const AggregateDataPtr& getNullKeyData() const { return null_key_data_; }

```

Comment:
What are the different user scenarios for above 4 methods? Seems a bit confusing.